crates/chain/src/persist.rs

```diff
@@ -0,0 +1,456 @@
+
+    /// Return the aggregate changeset `C` from persistence.
+    fn load_changes(&mut self) -> Result<Option<C>, Self::LoadError>;
+}
+
+impl<C> PersistBackend<C> for () {
+    type WriteError = Infallible;
+    type LoadError = Infallible;
+
+    fn write_changes(&mut self, _changeset: &C) -> Result<(), Self::WriteError> {
+        Ok(())
+    }
+
+    fn load_changes(&mut self) -> Result<Option<C>, Self::LoadError> {
+        Ok(None)
+    }
+}
+
+#[cfg(feature = "async")]
+/// An async persistence backend for writing and loading changesets.
+///
+/// `C` represents the changeset; a datatype that records changes made to in-memory data structures
+/// that are to be persisted, or retrieved from persistence.
+#[async_trait]
+pub trait PersistBackendAsync<C> {
+    /// The error the backend returns when it fails to write.
+    type WriteError: Debug + Display;
+
+    /// The error the backend returns when it fails to load changesets `C`.
+    type LoadError: Debug + Display;
+
+    /// Writes a changeset to the persistence backend.
+    ///
+    /// It is up to the backend what it does with this. It could store every changeset in a list or
+    /// it inserts the actual changes into a more structured database. All it needs to guarantee is
+    /// that [`load_from_persistence`] restores a keychain tracker to what it should be if all
+    /// changesets had been applied sequentially.
+    ///
+    /// [`load_from_persistence`]: Self::load_changes
+    async fn write_changes(&mut self, changeset: &C) -> Result<(), Self::WriteError>;
+
+    /// Return the aggregate changeset `C` from persistence.
+    async fn load_changes(&mut self) -> Result<Option<C>, Self::LoadError>;
+}
+
+#[cfg(feature = "async")]
+#[async_trait]
+impl<C> PersistBackendAsync<C> for () {
+    type WriteError = Infallible;
+    type LoadError = Infallible;
+
+    async fn write_changes(&mut self, _changeset: &C) -> Result<(), Self::WriteError> {
+        Ok(())
+    }
+
+    async fn load_changes(&mut self) -> Result<Option<C>, Self::LoadError> {
+        Ok(None)
+    }
+}
+
+/// `StagedPersistBackend` adds a convenient staging area for changesets before they are persisted.
+///
+/// Not all changes to the in-memory representation needs to be written to disk right away, so
+/// [`crate::persist::StagedPersistBackend::stage`] can be used to *stage* changes first and then
+/// [`crate::persist::StagedPersistBackend::commit`] can be used to write changes to disk.
+pub struct StagedPersistBackend<C, P: PersistBackend<C>> {
+    inner: P,
+    stage: C,
+}
+
+impl<C, P: PersistBackend<C>> PersistBackend<C> for StagedPersistBackend<C, P> {
+    type WriteError = P::WriteError;
+    type LoadError = P::LoadError;
+
+    fn write_changes(&mut self, changeset: &C) -> Result<(), Self::WriteError> {
+        self.inner.write_changes(changeset)
+    }
+
+    fn load_changes(&mut self) -> Result<Option<C>, Self::LoadError> {
+        self.inner.load_changes()
+    }
+}
```

Comment:
I don't think the staged area should impl `PersistBackend`. The `load_changes` method would not load everything pushed in `write_changes`. I guess we can change `load_changes` to merge on whatever is in `stage`, but I don't think that is helpful?

Reply:
It no longer makes sense to have `StagedPersistBackend` since we don't use it! The examples can do what we do in wallet.

Reply:
Change of plan (as discussed in DMs). The main selling point of `StagedPersist..` is in the `commit`-esc methods. Since the `commit` method checks whether it's an empty changeset (does not persist empty changeset) and not taking from the staged error if writing has failed. We should still keep this, but have `StagedPersist::inner` as in external input (so we can use it in wallet).